### PR TITLE
Added support for linux sockaddr_pppox.

### PR DESCRIPTION
**Pull request by alvardes**

### CODE REVIEW

**Comment by JohnTitor** [comment]
Hm, CI fails. I think the definition is wrong.

**Comment by JohnTitor** [comment]
Note: It seems you failed to rebase (given from CI).

**Comment by alvardes** [comment]
Hmm... 

@JohnTitor Can you help me?

Why does this fail on mips and sparc? It seems that the layout is the same. The roundtrip test however fails :\ Any idea?

**Comment by JohnTitor** [comment]
So, at a glance, `#[repr(packed)]` causes the failure, I guess?

**Comment by alvardes** [comment]
@JohnTitor Could you take a look at this:

https://dev.azure.com/rust-lang2/libc/_build/results?buildId=1917&view=logs&j=c8d582de-107e-5fbd-122b-5384eb107855&t=7bb3b962-ee23-58e7-68da-769fcc260c0d&l=2371

```
o you wish to create a custom hardware profile? [no] 
Error processing tar file(exit status 1): write /root/.android/avd/aarch64.avd/userdata.img: no space left on device

##[error]Bash exited with code '1'.

```

**Comment by JohnTitor** [comment]
I guess it's temporary, I saw that failure on rust-lang/rust as well.
Let me close/re-open to re-trigger CI..

**Comment by JohnTitor** [comment]
Ugh, it still appears. Then my another guess is that our cache storage is full.
cc @pietroalbini thoughts --^?

**Comment by alvardes** [comment]
@JohnTitor 

Now it passes all the test except `aarch64-unknown-linux-android`. 

What now?

(Because 1.13 doesn't support unions I only added support for `sockaddr_pppox` above 1.13. `pppoe_addr` and `pptp_addr` which are simple structs are added for all versions. Is that a good or should I cfg them too so that they are only present above 1.13?)

File: libc-test/build.rs [hunk 1]
```diff
@@ -2231,6 +2231,9 @@ fn test_linux(target: &str) {
     let aarch64_musl = target.contains("aarch64") && musl;
     let gnuabihf = target.contains("gnueabihf");
 
+    let supports_unaligned_memory_access =
+        !(mips32 || mips64 || sparc64 || s390x);
```
Comment (JohnTitor) [inline libc-test/build.rs:2235]: Should we really skip tests on their targets?

File: src/unix/linux_like/linux/mod.rs [hunk 2]
```diff
@@ -504,6 +504,33 @@ s! {
...
+        pub dev: [u8; 16]
+    }
+
+    pub struct pptp_addr {
+        pub call_id: u16,
+        pub sin_addr: ::in_addr
+    }
+}
+
+#[cfg(libc_union)]
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:520]: We could use `cfg_if!` macro.

File: src/unix/linux_like/linux/mod.rs [hunk 3]
```diff
@@ -597,6 +624,97 @@ cfg_if! {
             }
         }
 
+        #[cfg(libc_union)]
+        #[allow(unused_unsafe)]
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:628]: I don't think we should allow it here.

File: src/unix/linux_like/linux/mod.rs [hunk 4]
```diff
@@ -597,6 +624,97 @@ cfg_if! {
...
+                    /* PX_PROTO_PPTP */
+                    2  => unsafe { self.sa_addr.pptp == other.sa_addr.pptp },
+
+                    _ => {
+                        debug_assert!(
+                            false,
+                            "The sockaddr_pppox only supports PX_PROTO_OE \
+                            and PX_PROTO_PPTP. sa_protocol was: {}",
+                            unsafe { self.sa_protocol.clone() }
+                        );
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:646]: This could be removed once it's ready to land.

File: src/unix/linux_like/linux/mod.rs [hunk 5]
```diff
@@ -597,6 +624,97 @@ cfg_if! {
...
+                    /* PX_PROTO_PPTP */
+                    2 => unsafe { self.sa_addr.pptp.hash(s) },
+
+                    _ => {
+                        debug_assert!(
+                            false,
+                            "The sockaddr_pppox only supports PX_PROTO_OE \
+                            and PX_PROTO_PPTP. sa_protocol was: {}",
+                            unsafe { self.sa_protocol.clone() }
+                        );
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:677]: Same here.

File: src/unix/linux_like/linux/mod.rs [hunk 6]
```diff
@@ -597,6 +624,97 @@ cfg_if! {
...
+                        (self.sa_family,
+                        self.sa_protocol,
+                        self.sa_addr.pptp).fmt(f)
+                    } else {
+                        debug_assert!(
+                            false,
+                            "The sockaddr_pppox only supports PX_PROTO_OE \
+                            and PX_PROTO_PPTP. sa_protocol was: {}",
+                            self.sa_protocol.clone()
+                        );
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:709]: Same here.

File: src/unix/linux_like/linux/mod.rs [hunk 7]
```diff
@@ -597,6 +624,97 @@ cfg_if! {
             }
         }
 
+        #[cfg(libc_union)]
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:627]: We could use `cfg_if!` macro here as well.

File: src/unix/linux_like/linux/mod.rs [hunk 8]
```diff
@@ -597,6 +624,97 @@ cfg_if! {
...
+            }
+        }
+
+        #[cfg(libc_union)]
+        impl ::fmt::Debug for sockaddr_pppox_sa_addr_t {
+            fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
+                assert_eq!(24, ::mem::size_of::<Self>());
+                unsafe { ::mem::transmute::<_, [u8; 24]> (*self).fmt(f) }
+            }
+        }
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:689]: Hm, I think we should the same way as well as other places.

**Comment by JohnTitor** [comment]
Also, CI issue is now fixed on master.

File: libc-test/build.rs (hunk 1 above)
Comment (alvardes) [reply libc-test/build.rs:2235]: The problem is that reading from packed fields should take into account that the field is unaligned. I don't think the rust side of the CTest generated code doesn't this, thats why the roundrip tests fail for the packed struct on these targets.

File: src/unix/linux_like/linux/mod.rs (hunk 8 above)
Comment (alvardes) [reply src/unix/linux_like/linux/mod.rs]: fixed

File: src/unix/linux_like/linux/mod.rs (hunk 7 above)
Comment (alvardes) [reply src/unix/linux_like/linux/mod.rs]: fixed

File: src/unix/linux_like/linux/mod.rs (hunk 6 above)
Comment (alvardes) [reply src/unix/linux_like/linux/mod.rs]: fixed

File: src/unix/linux_like/linux/mod.rs (hunk 5 above)
Comment (alvardes) [reply src/unix/linux_like/linux/mod.rs]: fixed

File: src/unix/linux_like/linux/mod.rs (hunk 4 above)
Comment (alvardes) [reply src/unix/linux_like/linux/mod.rs]: fixed

File: src/unix/linux_like/linux/mod.rs (hunk 3 above)
Comment (alvardes) [reply src/unix/linux_like/linux/mod.rs]: fixed

File: src/unix/linux_like/linux/mod.rs (hunk 2 above)
Comment (alvardes) [reply src/unix/linux_like/linux/mod.rs:520]: fixed

File: libc-test/build.rs (hunk 1 above)
Comment (JohnTitor) [reply libc-test/build.rs:2235]: So, I think it's better to have different definitions. Did you confirm that items work well on skipped targets?

**Comment by JohnTitor** [comment]
Oh sorry, I somehow overlooked this one! If you're still interested, it'd be great if you could resolve merge conflicts. I'd like to see the CI result and take a look at the changes once it's done.

**Comment by JohnTitor** [comment]
I'm going to close as inactive, feel free to rebase and re-open when you have some time to work on.

**Pull request closed without merging**